### PR TITLE
Fixes all duplicate scoped synonym

### LINKED ISSUE #2305: [Typo/Error] Synonym appears as both exact AND broad or related for the same term
**Uberon term**
Multiple terms.
For example, https://www.ebi.ac.uk/ols/ontologies/uberon/terms?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FUBERON_0013528 (and other Brodmann area terms).

**Bug description**
There are several instances where the same synonym is used as both an exact synonym and a broad synonym for the same term. There are also instances where a synonym is annotated as both an exact and related synonym. For example, see UBERON_0013529 Brodmann (1909) area 11:

Screenshot 2022-02-14 at 09 32 34

In this case, the repeated synonyms probably should only be related to the term with the has_exact_synonym relationship.

### PR DESCRIPTION
**Pull request by anitacaron**
Two cases were fixed, when:
1. same term label and any other scoped synonym label
2. same label in two different scoped synonym

The cases were fixed to have `oboInOwl:hasExactSynonym`.

The following sparql update query were used (here it's mixed the fixes for the two cases):

```
PREFIX owl: <http://www.w3.org/2002/07/owl#>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX oboInOwl: <http://www.geneontology.org/formats/oboInOwl#>

DELETE {
 ?cls ?synonym ?value .
    ?x a owl:Axiom ;
           owl:annotatedSource ?cls ;
           owl:annotatedProperty ?synonym ;
           owl:annotatedTarget ?value ;
           ?p ?z .
}
INSERT {
 ?cls oboInOwl:hasExactSynonym ?value .
    ?x a owl:Axiom ;
           owl:annotatedSource ?cls ;
           owl:annotatedProperty oboInOwl:hasExactSynonym ;
           owl:annotatedTarget ?value ;
           ?p ?z .
}
WHERE {
  VALUES ?synonym {oboInOwl:hasExactSynonym oboInOwl:hasRelatedSynonym oboInOwl:hasNarrowSynonym oboInOwl:hasBroadSynonym} 
VALUES ?synonym2 {oboInOwl:hasExactSynonym oboInOwl:hasRelatedSynonym oboInOwl:hasNarrowSynonym oboInOwl:hasBroadSynonym} 
  ?cls rdfs:label ?label .
  ?cls ?synonym ?value .
  ?cls ?synonym2 ?value .
  OPTIONAL {
    ?x a owl:Axiom ;
           owl:annotatedSource ?cls ;
           owl:annotatedProperty ?synonym ;
           owl:annotatedTarget ?value ;
           ?p ?z .
  }
  FILTER (str(?label)=str(?value))
  FILTER(?synonym != ?synonym2)
  FILTER( !isBlank(?cls) && regex(str(?cls), "^http://purl.obolibrary.org/obo/UBERON_"))
}
```

Fixes #2305

### CODE REVIEW

**Comment by shawntanzk** [comment]
converting to a draft so it doesn't accidentally get merged - this needs to be reviewed before merging. I'll post a diff here to make reviewing easier.

**Comment by shawntanzk** [comment]
Diff too big to post here, please see: 
https://drive.google.com/file/d/1vc-uRZ7A6XnsMAsB05UGXRE7vO-5y8gf/view?usp=sharing

**Comment by matentzn** [comment]
@anitacaron can you check in the update queries(src/sparql/update)? We should do that to make it easier to do similar things in the future. Also, I think we should probably split both problems into separate update queries..

I think you don't capture case 1 entirely because you only look at terms that have at least two synonyms?

**Comment by anitacaron** [comment]
@matentzn I used two queries to make the changes. I put here the two mixed because the difference is a few lines.

File: src/sparql/update/remove-duplicate-synonyms.ru [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+      owl:annotatedSource ?cls ;
+      owl:annotatedProperty ?synonym ;
+      owl:annotatedTarget ?value ;
+      ?p ?z .
+}
+INSERT {
+ ?cls oboInOwl:hasExactSynonym ?value .
+    ?x a owl:Axiom ;
+      owl:annotatedSource ?cls ;
+      owl:annotatedProperty oboInOwl:hasExactSynonym ;
```
Comment (matentzn) [inline src/sparql/update/remove-duplicate-synonyms.ru:18]: This turns every synonym regardless of what it originally was to exact. In particular, if synonym 1 is broad, and synonym 2 is narrow, both will be turned into exact.

File: src/sparql/update/remove-duplicate-synonyms.ru [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      owl:annotatedTarget ?value ;
+      ?p ?z .
+}
+WHERE {
+  VALUES ?synonym {oboInOwl:hasExactSynonym oboInOwl:hasRelatedSynonym oboInOwl:hasNarrowSynonym oboInOwl:hasBroadSynonym}
```
Comment (matentzn) [inline src/sparql/update/remove-duplicate-synonyms.ru:23]: I recommend changing this to

```
  VALUES ?synonym { oboInOwl:hasRelatedSynonym oboInOwl:hasNarrowSynonym oboInOwl:hasBroadSynonym} 
  VALUES ?synonym2 {oboInOwl:hasExactSynonym } 
```

This ensures that at least one of the two baldy scoped synonyms is exact before assuming they both are.

File: src/sparql/update/remove-duplicate-synonyms.ru (hunk 1 above)
Comment (anitacaron) [reply src/sparql/update/remove-duplicate-synonyms.ru:18]: Yes, but it only changes the cases when they have the same value.

```
25  ?cls ?synonym ?value .
26  ?cls ?synonym2 ?value .
```

This decision came from the meeting https://github.com/obophenotype/uberon/issues/2305#issuecomment-1054564265

Comment (matentzn) [reply src/sparql/update/remove-duplicate-synonyms.ru:18]: This seems like a very odd decision to make. Imagine the case oboInOwl:hasRelatedSynonym oboInOwl:hasNarrowSynonym - it seems wrong to change them both to exact!

Comment (anitacaron) [reply src/sparql/update/remove-duplicate-synonyms.ru:18]: Ah, understood now. In this case, it still has some problems with duplicate scoped synonyms.

Comment (matentzn) [reply src/sparql/update/remove-duplicate-synonyms.ru:18]: How many? Maybe we can fix them manually.

Comment (anitacaron) [reply src/sparql/update/remove-duplicate-synonyms.ru:18]: It's about 15 cases.

File: src/sparql/update/remove-scoped-synonym-label.ru [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  ?cls rdfs:label ?label .
+  ?cls ?synonym ?value .
+  OPTIONAL {
+    ?x a owl:Axiom ;
+      owl:annotatedSource ?cls ;
+      owl:annotatedProperty ?synonym ;
+      owl:annotatedTarget ?value ;
+      ?p ?z .
+  }
+  FILTER(STR(?label) = STR(?value))
```
[the remark this answers is not in the document]
Comment (anitacaron) [reply src/sparql/update/remove-scoped-synonym-label.ru:33]: No strong reason, just clarity. I'll put the same variable name.

File: src/sparql/update/remove-duplicate-synonyms.ru (hunk 2 above)
Comment (anitacaron) [reply src/sparql/update/remove-duplicate-synonyms.ru:23]: done

File: src/sparql/update/remove-duplicate-synonyms.ru (hunk 1 above)
Comment (matentzn) [reply src/sparql/update/remove-duplicate-synonyms.ru:18]: Cool. Would you be ok to fix them manually?

broad/related --> broad
narrow/related --> narrow
broad/narrow --> try to understand, if unclear: close

Comment (shawntanzk) [reply src/sparql/update/remove-duplicate-synonyms.ru:18]: if its 15 cases, I would rather fix them manually on a case to case basis. Happy to look into it if you open a ticket and list these 15 terms down - happy for you to assign me :) thanks

Comment (anitacaron) [reply src/sparql/update/remove-duplicate-synonyms.ru:18]: no worries, Shawn. I did the modifications in this commit https://github.com/obophenotype/uberon/pull/2332/commits/0d0530a0e10798f3230969d1180daf6ec8f79830

Comment (shawntanzk) [reply src/sparql/update/remove-duplicate-synonyms.ru:18]: will run a robot diff again and get it ready for review :)

Comment (anitacaron) [reply src/sparql/update/remove-duplicate-synonyms.ru:18]: it's better if you check the diff in Git Desktop so you have the context 😄

Comment (shawntanzk) [reply src/sparql/update/remove-duplicate-synonyms.ru:18]: ah wanted to do the diff in robot diff cause its super big changes, but it seems like you can actually let the diff load here. It's mainly for others to confirm that the exacts are actually exact synonyms and no mistakes were made, but totally get how its actually easier to see here than in robot diff at times

**Comment by shawntanzk** [comment]
ROBOT diff: https://drive.google.com/file/d/1T9Zt02NLEFUhOaHYNcK7MeInFNk6xVzy/view?usp=sharing

File: src/sparql/update/remove-duplicate-synonyms.ru (hunk 1 above)
Comment (anitacaron) [reply src/sparql/update/remove-duplicate-synonyms.ru:18]: No matter how big the diffs are, you can always open them on your machine using Git Desktop. Here on the web, that is not possible.

The commit I commented here has only the 15 cases with duplicated synonym scope that are not exact; that's why you can see them.

**Comment by shawntanzk** [comment]
Notice that this adds an additional has_exact_synoym. Can it instead just remove it but move the dbxref annotation to the exact synonym (not sure how difficult that is)
Screenshot 2022-03-03 at 13 17 13

File: src/sparql/update/remove-duplicate-synonyms.ru (hunk 1 above)
Comment (shawntanzk) [reply src/sparql/update/remove-duplicate-synonyms.ru:18]: Oh i meant i wanted to see the final product compared to master to see how this PR would change the ontology

**Comment by matentzn** [comment]
@shawntanzk dont worry about that, we can use normalisation for contracting these duplicate synonyms! No need to change the SPARQL I think.

**Review by matentzn — APPROVED** [review]

**Comment by matentzn** [comment]
Great job! Assuming this is done?

**Comment by shawntanzk** [comment]
> Assuming this is done?

In the call, there was an agreement that this needs to be reviewed, cant rmbr by who - @dosumis @addiehl? will also look through it tmr :)

**Comment by matentzn** [comment]
I think the only thing that _can_ be reviewed is wether the synonyms are all exact. Many are not. But this is not really a hill you want to die on here. It will take dozens of curation hours to fix something that has been like this for 15 years, and no one complained so far. I would not get bogged down by details and just focus on QC and infrastructure so we can actually ensure that stuff that is added moving forward is not broken.

**Comment by matentzn** [comment]
@shawntanzk can we merge this? Seems a waste to let this open for people to never review this..

**Comment by shawntanzk** [comment]
honestly not super sure, would rather bring this up in a call - people had quite strong opinions about this last time it was discussed

**Comment by matentzn** [comment]
- [ ] Blocked by: https://github.com/obophenotype/uberon/issues/2363

**Comment by anitacaron** [comment]
I'm closing this PR as it's already too old, and it's better to have a new, clean PR. I've linked the queries used here in the issue.

**Pull request closed without merging**